payu/models/cesm_cmeps.py

```diff
@@ -156,15 +154,50 @@ def setup(self):
 
         self.runconfig.set("ALLCOMP_attributes", "start_type", start_type)
 
-        # Check pelayout makes sense
-        all_realms = self.realms + ["glc", "lnd"]
+        # check pelayout fits within requested cpucount
         cpucount = int(
             self.expt.config.get('ncpus', multiprocessing.cpu_count())
             )
+        all_realms = self.realms + ["glc", "lnd"]
         for realm in all_realms:
             ntasks = int(self.runconfig.get("PELAYOUT_attributes", f"{realm}_ntasks"))
-            assert cpucount >= ntasks, "Insufficient cpus for the pelayout in nuopc.runconfig"
-        
+            rootpe = int(self.runconfig.get("PELAYOUT_attributes", f"{realm}_rootpe"))
+            pestride = int(self.runconfig.get("PELAYOUT_attributes", f"{realm}_pestride"))
+            # rootpe is zero-based
+            if cpucount < (rootpe + ntasks*pestride):
+                raise ValueError(
+                    f"Insufficient cpus for the {realm} pelayout in nuopc.runconfig"
+                )
+
+        # check iolayout
+        for realm in self.realms:
+            # med and cpl names are both used in runconfig
+            if realm == "cpl": 
+                realm = "MED"
+            io_section = f"{realm.upper()}_modelio"
+            nc_type = self.runconfig.get(io_section, "pio_typename")
+            if nc_type == "netcdf4c":
+                raise ValueError(
+                    f"netcdf4c in {io_section} of nuopc.runconfig is deprecated, "
+                      "use netcdf4p"
+                )
+            else:
+                # if nc_type is netcdf, only one pe is needed
+                ioroot = int(self.runconfig.get(io_section, "pio_root"))
+                if cpucount < int(ioroot):
+                    raise ValueError(
+                        f"Insufficient cpus for the {io_section} ioroot pe in "
+                        "nuopc.runconfig"
+                    )
+                if nc_type == "netcdf4p":
+                    niotasks = int(self.runconfig.get(io_section, "pio_numiotasks"))
+                    iostride = int(self.runconfig.get(io_section, "pio_stride"))
+                    if cpucount <= (ioroot + niotasks*iostride):
+                        raise ValueError(
+                            f"The iolayout for {io_section} in nuopc.runconfig is "
+                            "requesting out of range cpus"
+                        )
```

Comment:
```suggestion
                assert ntasks * pestride >= niotasks * \
                    iostride, (
                        f"The iolayout for {io_section} in nuopc.runconfig is requesting "
                        f"out of range cpus of {realm}_component ({ntasks*pestride})"
                    )
```

Instead of checking the total `ncpus`, here should be on the number of cpus for each component - `ntasks * pestride`

Reply:
I sort of agree. I think we should be comparing to {realm}_ntasks rather than cpucoun